lib/services/warning-service.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * @fileoverview Emits warnings for ESLint.
+ * @author Francesco Trotta
+ */
+
+"use strict";
+
+//-----------------------------------------------------------------------------
+// Exports
+//-----------------------------------------------------------------------------
+
+/**
+ * A service that emits warnings ESLint.
+ */
+class WarningService {
+	/**
+	 * A function called internally to emit warnings using the Node.js API, if available.
+	 * @type {((warning: string, type: string) => void) | undefined}
+	 */
+	emitWarning = globalThis.process?.emitWarning;
```

Comment:
If `globalThis.process` is undefined then `emitWarning` will also be undefined, which will cause errors when any method is called. Should we provide a backup?
```suggestion
	emitWarning = globalThis.process?.emitWarning ?? () => {};
```

Reply:
For testing purposes, it might actually be better to define a constructor for this class so we can easily override where it gets `emitWarning`. Something like this?

```js
constructor({ systemEmitWarning = globalThis.process?.emitWarning ?? () => {}) {
   this.emitWarning = systemEmitWarning;
}
```

Reply:
I wrote unit tests to ensure that WarningService methods used by the Linter do not throw errors when `process` is undefined, e.g. in a browser:

https://github.com/eslint/eslint/blob/ee800528f553b8aa5d7e54ea3dd557fda88c57fd/tests/lib/services/warning-service.js#L103-L129

I will add a constructor to set `emitWarning` with a fallback to an empty function when `process` is undefined.